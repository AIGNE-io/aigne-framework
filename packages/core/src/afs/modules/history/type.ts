import type { AFSEntry } from "../../type.js";

export interface HistoryStorageCreateEntry extends Omit<AFSEntry, "id"> {}

export interface HistoryStorageListOptions {
  filter?: {
    userId?: string;
    sessionId?: string;
  };
  limit?: number;
  orderBy?: [string, "asc" | "desc"][];
}

export abstract class HistoryStorage {
  abstract create(entry: HistoryStorageCreateEntry): Promise<AFSEntry>;

  abstract list(options?: HistoryStorageListOptions): Promise<{ list: AFSEntry[] }>;

  abstract read(path: string): Promise<AFSEntry | undefined>;
}
