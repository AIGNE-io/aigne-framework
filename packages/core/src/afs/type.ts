import type { Emitter } from "strict-event-emitter";
import type { Context, ContextEventMap } from "../aigne/context.js";
import type { AFSStorage } from "./storage/type.js";

export interface AFSListOptions {
  filter?: {
    userId?: string;
    sessionId?: string;
  };
  recursive?: boolean;
  maxDepth?: number;
  limit?: number;
  orderBy?: [string, "asc" | "desc"][];
}

export interface AFSSearchOptions {
  limit?: number;
}

export interface AFSWriteEntryPayload extends Omit<AFSEntry, "id" | "path"> {}

export interface AFSModule {
  readonly moduleId: string;

  readonly path: string;

  onMount?(root: AFSRoot): void;

  list?(path: string, options?: AFSListOptions): Promise<{ list: AFSEntry[] }>;

  read?(path: string): Promise<AFSEntry | undefined>;

  write?(path: string, content: AFSWriteEntryPayload): Promise<AFSEntry>;

  search?(path: string, query: string, options?: AFSSearchOptions): Promise<{ list: AFSEntry[] }>;
}

export type AFSRootEvents = ContextEventMap & {
  historyCreated: [{ context: Context; entry: AFSEntry }];
};

export interface AFSRoot extends Emitter<AFSRootEvents>, AFSModule {
  storage(module: AFSModule): AFSStorage;
}

export interface AFSEntry<T = any> {
  id: string;
  createdAt?: Date;
  updatedAt?: Date;
  path: string;
  userId?: string | null;
  sessionId?: string | null;
  summary?: string | null;
  metadata?: Record<string, any> | null;
  linkTo?: string | null;
  content?: T;
}
