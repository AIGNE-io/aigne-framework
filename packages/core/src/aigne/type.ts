import type { Agent } from "../agents/agent.ts";

export interface AIGNECLIAgents {
  chat?: Agent;

  agents?: AIGNECLIAgent[];
}

export interface AIGNECLIAgent {
  agent?: Agent;
  name?: string;
  alias?: string[];
  description?: string;
  commands?: AIGNECLIAgent[];
}
