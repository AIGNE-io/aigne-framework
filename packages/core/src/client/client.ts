import type {
  AgentCallOptions,
  AgentResponse,
  AgentResponseChunk,
  AgentResponseStream,
  Message,
} from "../agents/agent.js";
import { AgentResponseStreamParser, EventStreamParser } from "../utils/event-stream.js";
import { tryOrThrow } from "../utils/type-utils.js";

export interface AIGNEClientOptions {
  url: string;
}

export class AIGNEClient {
  constructor(public options: AIGNEClientOptions) {}

  async call<I extends Message, O extends Message>(
    agent: string,
    input: I,
    options: AgentCallOptions & { streaming: true },
  ): Promise<AgentResponseStream<O>>;
  async call<I extends Message, O extends Message>(
    agent: string,
    input: I,
    options?: AgentCallOptions & { streaming?: false },
  ): Promise<O>;
  async call<I extends Message, O extends Message>(
    agent: string,
    input: I,
    options?: AgentCallOptions,
  ): Promise<AgentResponse<O>>;
  async call<I extends Message, O extends Message>(
    agent: string,
    input: I,
    options?: AgentCallOptions,
  ): Promise<AgentResponse<O>> {
    const response = await this.fetch(this.options.url, {
      method: "POST",
      headers: {
        "Content-Type": "application/json",
      },
      body: JSON.stringify({ agent, input, options }),
    });

    if (!options?.streaming) {
      return await response.json();
    }

    const stream = response.body;
    if (!stream) throw new Error("Response body is not a stream");

    return stream
      .pipeThrough(new TextDecoderStream())
      .pipeThrough(new EventStreamParser<AgentResponseChunk<O>>())
      .pipeThrough(new AgentResponseStreamParser());
  }

  async fetch(...args: Parameters<typeof globalThis.fetch>): Promise<Response> {
    const result = await globalThis.fetch(...args);

    if (!result.ok) {
      let message: string | undefined;

      try {
        const text = await result.text();
        const json = tryOrThrow(() => JSON.parse(text) as { error?: { message: string } });
        message = json?.error?.message || text;
      } catch {
        // ignore
      }

      throw new Error(`Failed to fetch url ${args[0]} with status ${result.status}: ${message}`);
    }

    return result;
  }
}
