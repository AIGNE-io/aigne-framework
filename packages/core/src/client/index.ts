/**
 * @module core/Client
 */

export * from "./client.js";
