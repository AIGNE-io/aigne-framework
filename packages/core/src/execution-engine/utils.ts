import type { FunctionAgentFn, Message } from "../agents/agent.js";
import type { Context, Runnable } from "./context.js";

export function sequential(..._agents: [Runnable, ...Runnable[]]): FunctionAgentFn {
  let agents = [..._agents];

  return async (input: Message, context?: Context) => {
    if (!context) throw new Error("Context is required for sequential agents");

    const output: Message = {};

    // Clone the agents to run, so that we can update the agents list during the loop
    const agentsToRun = [...agents];
    agents = [];

    for (const agent of agentsToRun) {
      const [o, transferToAgent] = await context.call(
        agent,
        { ...input, ...output },
        { returnActiveAgent: true },
      );
      Object.assign(output, o);
      agents.push(transferToAgent);
    }

    return output;
  };
}

export function parallel(..._agents: [Runnable, ...Runnable[]]): FunctionAgentFn {
  let agents = [..._agents];

  return async (input: Message, context?: Context) => {
    if (!context) throw new Error("Context is required for parallel agents");

    const result = await Promise.all(
      agents.map((agent) => context.call(agent, input, { returnActiveAgent: true })),
    );

    agents = result.map((i) => i[1]);
    const outputs = result.map((i) => i[0]);

    return Object.assign({}, ...outputs);
  };
}
