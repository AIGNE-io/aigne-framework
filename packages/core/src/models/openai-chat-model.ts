import { nanoid } from "nanoid";
import OpenAI from "openai";
import type { ChatCompletionMessageParam, ChatCompletionTool } from "openai/resources";
import { parseJSON } from "../utils/json-schema.js";
import { isNonNullable } from "../utils/type-utils.js";
import {
  ChatModel,
  type ChatModelInput,
  type ChatModelInputMessage,
  type ChatModelInputTool,
  type ChatModelOptions,
  type ChatModelOutput,
  type Role,
} from "./chat-model.js";

const CHAT_MODEL_OPENAI_DEFAULT_MODEL = "gpt-4o-mini";

export interface OpenAIChatModelOptions {
  apiKey?: string;
  baseURL?: string;
  model?: string;
  modelOptions?: ChatModelOptions;
}

export class OpenAIChatModel extends ChatModel {
  constructor(public options?: OpenAIChatModelOptions) {
    super();
  }

  private _client?: OpenAI;

  private get client() {
    if (!this.options?.apiKey) throw new Error("Api Key is required for OpenAIChatModel");

    this._client ??= new OpenAI({
      baseURL: this.options.baseURL,
      apiKey: this.options.apiKey,
    });
    return this._client;
  }

  get modelOptions() {
    return this.options?.modelOptions;
  }

  async process(input: ChatModelInput): Promise<ChatModelOutput> {
    const res = await this.client.chat.completions.create({
      model: this.options?.model || CHAT_MODEL_OPENAI_DEFAULT_MODEL,
      temperature: input.modelOptions?.temperature ?? this.modelOptions?.temperature,
      top_p: input.modelOptions?.topP ?? this.modelOptions?.topP,
      frequency_penalty:
        input.modelOptions?.frequencyPenalty ?? this.modelOptions?.frequencyPenalty,
      presence_penalty: input.modelOptions?.presencePenalty ?? this.modelOptions?.presencePenalty,
      messages: await contentsFromInputMessages(input.messages),
      tools: toolsFromInputTools(input.tools),
      tool_choice: input.toolChoice,
      parallel_tool_calls: !input.tools?.length
        ? undefined
        : (input.modelOptions?.parallelToolCalls ?? this.modelOptions?.parallelToolCalls),
      response_format:
        input.responseFormat?.type === "json_schema"
          ? {
              type: "json_schema",
              json_schema: {
                ...input.responseFormat.jsonSchema,
                schema: jsonSchemaToOpenAIJsonSchema(input.responseFormat.jsonSchema.schema),
              },
            }
          : undefined,
      stream: true,
    });

    let text = "";
    const toolCalls: (NonNullable<ChatModelOutput["toolCalls"]>[number] & {
      args: string;
    })[] = [];

    for await (const chunk of res) {
      const choice = chunk.choices?.[0];

      if (choice?.delta.tool_calls?.length) {
        for (const call of choice.delta.tool_calls) {
          toolCalls[call.index] ??= {
            id: call.id || nanoid(),
            type: "function" as const,
            function: { name: "", arguments: {} },
            args: "",
          };
          const c = toolCalls[call.index];
          if (!c) throw new Error("Tool call not found");

          if (call.type) c.type = call.type;

          c.function.name = c.function.name + (call.function?.name || "");
          c.args = c.args.concat(call.function?.arguments || "");
        }
      }

      if (choice?.delta.content) text += choice.delta.content;
    }

    const result: ChatModelOutput = {};

    if (input.responseFormat?.type === "json_schema" && text) {
      result.json = parseJSON(text);
    } else {
      result.text = text;
    }

    if (toolCalls.length) {
      result.toolCalls = toolCalls.map(({ args, ...c }) => ({
        ...c,
        function: { ...c.function, arguments: parseJSON(args) },
      }));
    }

    return result;
  }
}

const ROLE_MAP: { [key in Role]: ChatCompletionMessageParam["role"] } = {
  system: "system",
  user: "user",
  agent: "assistant",
  tool: "tool",
} as const;

async function contentsFromInputMessages(
  messages: ChatModelInputMessage[],
): Promise<ChatCompletionMessageParam[]> {
  return messages.map(
    (i) =>
      ({
        role: ROLE_MAP[i.role],
        content:
          typeof i.content === "string"
            ? i.content
            : i.content
                ?.map((c) => {
                  if (c.type === "text") {
                    return { type: "text" as const, text: c.text };
                  }
                  if (c.type === "image_url") {
                    return {
                      type: "image_url" as const,
                      image_url: { url: c.url },
                    };
                  }
                })
                .filter(isNonNullable),
        tool_calls: i.toolCalls?.map((i) => ({
          ...i,
          function: {
            ...i.function,
            arguments: JSON.stringify(i.function.arguments),
          },
        })),
        tool_call_id: i.toolCallId,
        name: i.name,
      }) as ChatCompletionMessageParam,
  );
}

function toolsFromInputTools(tools?: ChatModelInputTool[]): ChatCompletionTool[] | undefined {
  return tools?.length
    ? tools.map((i) => ({
        type: "function",
        function: {
          name: i.function.name,
          description: i.function.description,
          parameters: i.function.parameters as Record<string, unknown>,
        },
      }))
    : undefined;
}

function jsonSchemaToOpenAIJsonSchema(schema: Record<string, unknown>): Record<string, unknown> {
  if (schema?.type === "object") {
    const { required, properties } = schema as {
      required?: string[];
      properties: Record<string, unknown>;
    };

    return {
      ...schema,
      properties: Object.fromEntries(
        Object.entries(properties || {}).map(([key, value]) => {
          const valueSchema = jsonSchemaToOpenAIJsonSchema(value as Record<string, unknown>);

          // NOTE: All fields must be required https://platform.openai.com/docs/guides/structured-outputs/all-fields-must-be-required
          return [
            key,
            required?.includes(key) ? valueSchema : { anyOf: [valueSchema, { type: ["null"] }] },
          ];
        }),
      ),
      required: Object.keys(properties || {}),
    };
  }

  if (schema?.type === "array") {
    const { items } = schema as { items: Record<string, unknown> };

    return {
      ...schema,
      items: jsonSchemaToOpenAIJsonSchema(items),
    };
  }

  return schema;
}
