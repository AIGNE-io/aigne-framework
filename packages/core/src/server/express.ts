import compression from "compression";
import express, { json, type RequestHandler } from "express";
import { z } from "zod";
import type { ExecutionEngine } from "../execution-engine/execution-engine.js";
import { readableStreamToAsyncIterator } from "../utils/stream-utils.js";
import { tryOrThrow } from "../utils/type-utils.js";
import { HttpError } from "./error.js";

export function createExpressMiddleware(engine: ExecutionEngine): RequestHandler {
  const router = express.Router();

  const bodySchema = z.object({
    agent: z.string(),
    input: z.record(z.string(), z.unknown()),
    options: z
      .object({
        streaming: z.boolean().nullish(),
      })
      .nullish(),
  });

  router.post("/", json(), compression(), async (req, res) => {
    try {
      const { input, options, ...body } = tryOrThrow(
        () => bodySchema.parse(req.body),
        (error) => new HttpError(400, error.message),
      );

      const agent = engine.agents[body.agent];
      if (!agent) throw new HttpError(404, `Agent ${body.agent} not found`);

      if (!options?.streaming) {
        const result = await engine.call(agent, input);
        res.json({ ...result });
        return;
      }

      const stream = await engine.call(agent, input, { streaming: true });

      res.setHeader("Content-Type", "text/event-stream");
      res.setHeader("Cache-Control", "no-cache");

      try {
        for await (const chunk of readableStreamToAsyncIterator(stream)) {
          res.write(`data: ${JSON.stringify(chunk)}\n\n`);
          res.flush();
        }
      } catch (error) {
        res.write(`event: error\ndata: ${JSON.stringify({ message: error.message })}\n\n`);
        res.flush();
      } finally {
        res.end();
      }
    } catch (error) {
      res
        .status(error instanceof HttpError ? error.status : 500)
        .json({ error: { message: error.message } });
    }
  });

  return router;
}
