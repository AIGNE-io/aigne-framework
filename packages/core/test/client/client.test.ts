import { afterEach, beforeEach, expect, spyOn, test } from "bun:test";
import assert from "node:assert";
import type { Server } from "node:http";
import { AIAgent, type ChatModel, ExecutionEngine, type Message } from "@aigne/core";
import { AIGNEClient } from "@aigne/core/client/client.js";
import { OpenAIChatModel } from "@aigne/core/models/openai-chat-model.js";
import { AIGNEServer } from "@aigne/core/server/server.js";
import { arrayToReadableStream, readableStreamToArray } from "@aigne/core/utils/stream-utils";
import { detect } from "detect-port";
import express from "express";

let port: number;
let model: ChatModel;
let httpServer: Server;
let url: string;

beforeEach(async () => {
  port = await detect();

  model = new OpenAIChatModel();

  const chat = AIAgent.from({
    name: "chat",
  });

  const engine = new ExecutionEngine({ model, agents: [chat] });

  const aigneServer = new AIGNEServer(engine);

  const server = express();

  server.use(express.json());

  server.post("/aigne/call", async (req, res) => {
    await aigneServer.call(req.body, res);
  });

  httpServer = server.listen(port);

  url = `http://localhost:${port}/aigne/call`;
});

afterEach(async () => {
  httpServer.closeAllConnections();
  httpServer.close();
});

test.each([{ streaming: false }, { streaming: true }])(
  "AIGNEClient should return correct response for %p",
  async (options) => {
    const client = new AIGNEClient({ url });
    const response = await client.call("chat", { $message: "hello" }, options);

    if (options.streaming) {
      assert(response instanceof ReadableStream);
      expect(readableStreamToArray(response)).resolves.toMatchSnapshot();
    } else {
      expect(response).toMatchSnapshot();
    }
  },
);

test.each([{ streaming: false }, { streaming: true }])(
  "AIGNEClient should return error response for %p",
  async (options) => {
    spyOn(model, "process").mockReturnValueOnce(
      Promise.resolve(
        arrayToReadableStream([
          { delta: { text: { text: "Hello" } } },
          { delta: { text: { text: ", world" } } },
          new Error("test error"),
        ]),
      ),
    );

    const client = new AIGNEClient({ url });
    const response = client.call("chat", { $message: "hello" }, options);

    if (options.streaming) {
      const stream = await response;
      assert(stream instanceof ReadableStream);
      expect(readableStreamToArray(stream, { catchError: true })).resolves.toMatchSnapshot();
    } else {
      expect(response).rejects.toThrow("test error");
    }
  },
);

test("AIGNEClient should return error response for not found agent", async () => {
  const client = new AIGNEClient({ url });

  const response = client.call("not-exists-agent", {});

  expect(response).rejects.toThrow("status 404: Agent not-exists-agent not found");
});

test("AIGNEClient should return error response for invalid request body", async () => {
  const client = new AIGNEClient({ url });

  const response = client.call("chat", [] as unknown as Message);

  expect(response).rejects.toThrow(
    "status 400: Call agent chat check arguments error: input: Expected object, received array",
  );
});
