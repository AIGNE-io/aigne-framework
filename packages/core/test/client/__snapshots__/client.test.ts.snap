// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`AIGNEClient should return correct response for {
  streaming: false,
} 1`] = `
{
  "$message": "Hello! How can I assist you today?",
}
`;

exports[`AIGNEClient should return correct response for {
  streaming: true,
} 1`] = `
[
  {
    "delta": {
      "json": {
        "$message": "Hello",
      },
      "text": {
        "$message": "Hello",
      },
    },
  },
  {
    "delta": {
      "json": {
        "$message": "Hello!",
      },
      "text": {
        "$message": "!",
      },
    },
  },
  {
    "delta": {
      "json": {
        "$message": "Hello! How",
      },
      "text": {
        "$message": " How",
      },
    },
  },
  {
    "delta": {
      "json": {
        "$message": "Hello! How can",
      },
      "text": {
        "$message": " can",
      },
    },
  },
  {
    "delta": {
      "json": {
        "$message": "Hello! How can I",
      },
      "text": {
        "$message": " I",
      },
    },
  },
  {
    "delta": {
      "json": {
        "$message": "Hello! How can I assist",
      },
      "text": {
        "$message": " assist",
      },
    },
  },
  {
    "delta": {
      "json": {
        "$message": "Hello! How can I assist you",
      },
      "text": {
        "$message": " you",
      },
    },
  },
  {
    "delta": {
      "json": {
        "$message": "Hello! How can I assist you today",
      },
      "text": {
        "$message": " today",
      },
    },
  },
  {
    "delta": {
      "json": {
        "$message": "Hello! How can I assist you today?",
      },
      "text": {
        "$message": "?",
      },
    },
  },
]
`;

exports[`AIGNEClient should return error response for {
  streaming: true,
} 1`] = `
[
  {
    "delta": {
      "json": {
        "$message": "Hello",
      },
      "text": {
        "$message": "Hello",
      },
    },
  },
  {
    "delta": {
      "json": {
        "$message": "Hello, world",
      },
      "text": {
        "$message": ", world",
      },
    },
  },
  [Error: test error],
]
`;
